sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -469,36 +583,121 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
+      }
+
+      if (getGqlQuery() != null && getGqlQuery().isAccessible()) {
+        checkNotNull(getGqlQuery().get(), "gqlQuery");
+      }
     }
 
     @Override
     public void populateDisplayData(DisplayData.Builder builder) {
       super.populateDisplayData(builder);
+      String query = getQuery() == null ? null : getQuery().toString();
       builder
           .addIfNotNull(DisplayData.item("projectId", getProjectId())
               .withLabel("ProjectId"))
           .addIfNotNull(DisplayData.item("namespace", getNamespace())
               .withLabel("Namespace"))
-          .addIfNotNull(DisplayData.item("query", getQuery().toString())
-              .withLabel("Query"));
+          .addIfNotNull(DisplayData.item("query", query)
+              .withLabel("Query"))
+          .addIfNotNull(DisplayData.item("gqlQuery", getGqlQuery()));
+    }
+
+    @VisibleForTesting
+    static class V1Options implements Serializable {
+      private final ValueProvider<String> project;
+      @Nullable
+      private final ValueProvider<String> namespace;
+      @Nullable
+      private final String localhost;
+
+      private V1Options(ValueProvider<String> project, ValueProvider<String> namespace,
+          String localhost) {
+        this.project = project;
+        this.namespace = namespace;
+        this.localhost = localhost;
+      }
+
+      public static V1Options from(String projectId, String namespace, String localhost) {
+        return from(StaticValueProvider.of(projectId), StaticValueProvider.of(namespace),
+            localhost);
+      }
+
+      public static V1Options from(ValueProvider<String> project, ValueProvider<String> namespace,
+          String localhost) {
+        return new V1Options(project, namespace, localhost);
+      }
+
+      public String getProjectId() {
+        return project.get();
+      }
+
+      @Nullable
+      public String getNamespace() {
+        return namespace == null ? null : namespace.get();
+      }
+
+      public ValueProvider<String> getProjectValueProvider() {
+        return project;
+      }
+
+      @Nullable
+      public ValueProvider<String> getNamespaceValueProvider() {
+        return namespace;
+      }
+
+      @Nullable
+      public String getLocalhost() {
+        return localhost;
+      }
+
     }
 
     /**
-     * A class for v1 Cloud Datastore related options.
+     * A DoFn that translates a Cloud Datastore gql query string to {@code Query}.
      */
-    @VisibleForTesting
-    @AutoValue
-    abstract static class V1Options implements Serializable {
-      public static V1Options from(String projectId, Query query, @Nullable String namespace,
-                                   @Nullable String localhost) {
-        return new AutoValue_DatastoreV1_Read_V1Options(projectId, query, namespace, localhost);
+    static class GqlQueryTranslateFn extends DoFn<ValueProvider<String>, Query> {
```

Comment:
Huh, is ValueProvider intended to be used as an element of a PCollection? I didn't know they are even encodable.